### PR TITLE
Add async stream return types

### LINKED ISSUE #16634: Missing C# 8 documentation
C# 8 adds an additional return type: IAsyncEnumerable&lt;T&gt; as long as you're using yield return instead of just return.

See the "What's new in C# 8" documentation under Asynchronous Streams: https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#asynchronous-streams

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d055727b-c97c-4d25-39b8-a4d3f21d0444
* Version Independent ID: c8a7c6f1-8ee1-5030-db12-3db65a968501
* Content: [Async Return Types (C#)](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/async-return-types#feedback)
* Content Source: [docs/csharp/programming-guide/concepts/async/async-return-types.md](https://github.com/dotnet/docs/blob/master/docs/csharp/programming-guide/concepts/async/async-return-types.md)
* Product: **dotnet-csharp**
* Technology: **csharp-async**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #16634 

This may be easier to review by individual commits:

1. Add that `IAsyncEnumerable<T>` is a valid return type for an async method. Add a new project and sample code for IAsyncEnumerable.
1. Fix style issues and remove unused anchors. (Those anchors were added during migration).
1. Moves the old snippet files without any changes.
1. Adapts the code to use the (new) preferred syntax. Update code to compile in a single project.

### CODE REVIEW

File: docs/csharp/programming-guide/concepts/async/snippets/async-returns1.cs [hunk 1]
```diff
@@ -30,9 +26,9 @@ static async Task<int> GetLeisureHours()
            leisureHours = 5;  
      
        return leisureHours;  
-   }  
+   }
+    // The example displays output like the following:
```
Comment (tdykstra) [inline docs/csharp/programming-guide/concepts/async/snippets/async-returns1.cs:30]: Are these lines deliberately indented one space?

File: docs/csharp/programming-guide/concepts/async/snippets/async-returns2.cs [hunk 2]
```diff
@@ -23,10 +19,11 @@ static async Task WaitAndApologize()
       // Task.Delay delays the following line by two seconds.  
       Console.WriteLine("\nSorry for the delay. . . .\n");  
    }
+    // The example displays the following output:
```
Comment (tdykstra) [inline docs/csharp/programming-guide/concepts/async/snippets/async-returns2.cs:22]: Are these lines deliberately indented one space?

File: docs/csharp/programming-guide/concepts/async/snippets/async-returns1a.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        int leisureHours;
+        if (today.First() == 'S')
+            leisureHours = 16;
+        else
+            leisureHours = 5;
+
+        return leisureHours;
+    }
+}
+// The example displays output like the following:
```
Comment (tdykstra) [inline docs/csharp/programming-guide/concepts/async/snippets/async-returns1a.cs:35]: The example as shown doesn't actually display anything, it returns that string.

**Review by tdykstra — APPROVED** [review]
Looks good, and the code all runs perfectly.

File: docs/csharp/programming-guide/concepts/async/async-return-types.md [hunk 4]
```diff
@@ -69,7 +66,15 @@ Because <xref:System.Threading.Tasks.Task> and <xref:System.Threading.Tasks.Task
 
 .NET provides the <xref:System.Threading.Tasks.ValueTask%601?displayProperty=nameWithType> structure as a lightweight implementation of a generalized task-returning value. To use the <xref:System.Threading.Tasks.ValueTask%601?displayProperty=nameWithType> type, you must add the `System.Threading.Tasks.Extensions` NuGet package to your project. The following example uses the <xref:System.Threading.Tasks.ValueTask%601> structure to retrieve the value of two dice rolls.
   
-[!code-csharp[return-value](../../../../../samples/snippets/csharp/programming-guide/async/async-valuetask.cs)]
+:::code language="csharp" source="./snippets/async-valuetask.cs":::
+
+## Async streams with IAsyncEnumerable\<T\>
+
+Starting with C# 8.0, an async method may return an *async stream*, represented by <xref:System.Collections.Generic.IAsyncEnumerable%601>. An async stream provides a way to enumerate a stream when elements are generated in chunks with repeated asynchronous calls. The following example shows an async method that generates an async stream:
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/csharp/programming-guide/concepts/async/async-return-types.md:73]: Good point. I changed to "enumerate items read from a stream" to be more clear.

**Pull request merged**